File: leetcode/21. Merge Two Sorted Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        ListNode head(0);
        ListNode* curr = &head;
        while(l1 != NULL || l2 != NULL){
            if(l1 == NULL){
                curr->next = l2;
                curr = l2;
                l2 = l2->next;
            } else if(l2 == NULL){
                curr->next = l1;
                curr = l1;
                l1 = l1->next;
            } else if(l2->val <= l1->val){
                curr->next = l2;
                curr = l2;
                l2 = l2->next;
            } else if(l1-> val < l2->val){
                curr->next = l1;
                curr = l1;
                l1 = l1->next;
            }
        }
        return head.next;
    }
};

File: leetcode/linkedlist/21. Merge Two Sorted Lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode()
        node = dummy
        node1 = list1
        node2 = list2
        while node1 is not None and node2 is not None:
            if node1.val < node2.val:
                node.next = node1
                node1 = node1.next
            else:
                node.next = node2
                node2 = node2.next
            node = node.next
        
        node1 = node2 if node2 is not None else node1
        while node1 is not None:
            node.next = node1
            node = node.next
            node1 = node1.next
        return dummy.next